### ISSUE TITLE
Browsers should go to an 'ydrv' sfs

### ISSUE THREAD

**Issue by 01micko** [puppylinux-woof-CE/woof-CE#639]
Usually, the first thing that needs updating in an OS is the browser. If the browser was sent to the 'adrv' then it wouldn't be part of the main sfs and could easily be updated with another adrv. Often people don't like the shipped browser and install another anyway.

This could be done in 3builddistro before the packages are copied to sandbox3/rootfs-complete. Of course it should be optional as well.

**Comment by mavrothal** [comment]
Is a good idea but adrv is loaded be default and is likely to have other apps too, so replacing is not going to be an easy user-task. Unless we have a or ydrv only for the browser, which is fine since they are not used at all right now and I do not see the "desktop" SFSs to appear anytime soon.

**Comment by 01micko** [comment]
Yeah, maybe ydrv then. I'll alter the title.

**Comment by dimkr** [comment]
IMHO there are too many *drv with this change. Maybe, we should change the SFS loading behavior so all SFSs called *drv_$DISTRO_FILE_PREFIX.sfs are loaded.

**Comment by q5sys** [comment]
I have to agree.  I understand the original reason for doing the different
'*drv', but I think going foward, Dima's suggestion seems to be the
simplest solution and also allows devs to order the sfs packages on boot of
they need to.  I know with TazOC's older Lighthouse versions, it was very
important that certain sfs were loaded first so that the versions of the
libs were proper as the layers were built up.
If we keep adding more '*drv' then it requires much more work to change the
structure as it becomes necessary in the future... effectively breaking
future backwards compatibility.

Btw, sorry for my absence of late, been dealing with things in the real
world.  Looking forward to get cracking on the official release of
slacko64.

**Comment by 01micko** [comment]
>IMHO there are too many *drv with this change. Maybe, we should change the SFS loading behavior so all SFSs called *drv_$DISTRO_FILE_PREFIX.sfs are loaded.

They are, except for `fdrv` in experimental fdrv branch. It's loaded with a boot parameter. There is no need for more. However we should consider bumping up the default number of sfs that can be loaded form 6 to something like 15 maybe?

**Comment by q5sys** [comment]
More than 6 would be helpful, I usually make a seperate sfs for browsers,
since they update so fast it makes it easier to update quickly.  Though
with a few browsers (FF, Chrome, Qupzilla)... I take up all my sfs slots.
I'm sort of crazy when it comes to sfs.  On Ligthouse I was using 16.  I
don't think I've gone above 20 except for once or twice in the past 5
years.  So anymore than that will rarely be useful.

**Comment by mavrothal** [comment]
For "user-choice" application-SFSs, I do not think we should mount them in the init like a,f,y,z-drv.
I also think that boot arguments is a "user unfriendly" "geeky" approach and should be used only as a last resort. Actually I believe that the minimal use of them is one of the strengths of puppy vs other small distros and should try to keep it that way.
We can still build and ship the browser as a separate SFS. Then  have defaultbrowser and the original browser menu entry, pointing to a script that will call sfs_load, mount the browser (and added to the extrasfs list for future boots) run the application and change the entries in menu and defaultbrowser accordingly. 
Adding another browser or replacing what is there should be done through a UI in the said (or an accompanying) script.

**Comment by technosaurus** [comment]
You guys are killing me with all this overcomplication.
I suggested this a long, long time ago, but here it goes again:
how about just using a simple boot parameter sfslist=
ex.
pfix=ram sfslist=devx.sfs,zdrv.sfs,firefox.sfs ...

then just use:
```bash
IFS=","
for SFS in $sfslist; do
load_sfs $SFS;
done
```
isn't the init script slow enough already without adding further complication?

**Comment by mavrothal** [comment]
I'm sure that init can be simpler but right now puppy loads a,z,ydrv without any boot parameter at the init and all the other SFSs as defined in BOOTCONFIG $extrafilelist. 
The only boot parameter is for the fdrv and could actually be eliminated and loaded by default if present.
The less boot parameters the better :smile:

**Comment by technosaurus** [comment]
@mavrothal - you are trading 1 powerful boot parameter for several lines of slow, clunky code though.  And maybe I  want to boot without my adrv or qdrv or whatever without having to move/delete the the damn thing.  

Maybe I want to try booting slacko with my debian kernel image and zdrv because it has the x32 ABI enabled.   But nope, can't do that either because we add a stupid ass chunk to the end of the files to compensate for the fact that we didn't know how to effectively use boot parameters.

In fact it could eliminate **"... searching for puppy files"** by allowing full paths like:
sfslist=/dev/sda1/slacko/zdrv.sfs,xxxx-xxxx-xxxx-xxxx/slacko/extras.sfs

Maybe we don't want to wait 10-30s for usb to load, we could just add an optional callback
sfslist=xxxx-xxxx-xxxx-xxxx/slacko/extras.sfs:slow_usb_callback

The same thing goes for the save file.  Using a savefile=... parameter allows you to make a different boot menu entry for each family member / student / customer /etc... (kinda useful since Puppy is single user)

The installer and init should be in the same package so that these types of things can be coordinated.

... but no less boot parameters are not necessarily better

**Comment by 01micko** [comment]
lets take a scenario...

Joe User has windows XP/Vista/7 and has managed to burn a puppy CD (he has an optical drive). It's his first experience with Linux. He's happy if he sees a desktop else that CD/DVD becomes a coaster or worse.

Then he needs network. If not coaster.

Now we have him. It's going to take him at least a month to realise that there are boot params. 

What we really need is a better utilisation of syslinux where we have a menu with:

- boot puppy
- boot puppy if graphics didn't work (ie: nomodeset)
- boot puppy without searching drives (aka pfix=ram)
- etcetera

? 

Of course that is the most simple situation. It can quickly become more complex, eg;

- no optical drive (how do we accommodate that? Maybe PUP2USB.BAT?
- UEFI (refind?)
- etcetera

yes we have work to do so I'll close this issue.

**Comment by mavrothal** [comment]
I have very little doubt that everything can be more efficient.
But puppy established a name by being user-friendly and comprehensive OOTB. Would be nice if we keep it that way. To that extend is important to do whatever needs to be done without forcing the user to do it for us (see boot parameters). 
Puppy can be installed in every imaginable way and in all kinds of media and partitions with files all over the place and still finds (slowly) its way through this mess. The real question in my mind at least is, can we keep doing all that the current puppy init does only much faster and efficient? I think yes, but first we need to identify all the different setups and then try to see how we could test for all different scenarios more efficiently.

**Comment by technosaurus** [comment]
@mavrothal - We don't _need_ to find anything.  The installer knows where it installs stuff.  All it has to do is put the paths in the sfslist, but instead it throws the baby out with the bathwater.

**Comment by peabee** [comment]
Not everybody uses the installer - I certainly don't!!

My vote is for keeping boot parameters simple for all the reasons mav 
and mick have stated.

**Comment by technosaurus** [comment]
@peabee - I don't either, but it could still be simple, just use some default like @dimkr suggested for manual installs
sfsfiles=*_$DISTRO_FILE_PREFIX.sfs
or even sfsfiles=/slacko64/*.sfs

... which would revert to the old slow way of finding each file in some random partition on some random drive in series by poking around on every single one of them and not bothering to check for the other sfs files on the same partition while it is mounted

Puppy shouldn't be poking around on drives that the user doesn't specify on a regular installation, just because its installer doesn't keep track.  For example, if I am going to use puppy to remove a virus on a computer's hard disk from a flash drive, it should never mount the infected internal drives.

**Comment by mavrothal** [comment]
I think that if we want puppy to do special things then is reasonable to have boot parameters.
But this is already happening. For example "pmedia=cd pfix=ram" does not look to any (infected ;) drives.
When booting from USB puppy does not look to internal HD by default. 
Defining pdev1=sdXY, only looks in sdXY. 
Defining directly pupsfs and a/z/y/f-drv (actually this is techno's code) does not even searches pdev1.
There is also an init patch in the forum (gyro?) to define pupsavefile (works fine when I tried it) so no need to search for this either.
We can further clean up/modify these parameters as/if needed, but even today we can make booting faster much faster with existing command line arguments.
What is really needed to succeed that is improve grub4dos configuration that is called after the installer, to pass the all these arguments to the boot command lines. At least when we are installing in the HD where the paths are known.
Bottom line (for me), if no argument is given we keep searching everything and find our files.
If the installer/grub4dos is used, we improve grub4dos configuration to pass *automatically* the correct arguments (pdev1, pupsfs, psavefile, sfslist, whatnot) and we just load our files without or minimal searching.
In every case we should NOT ask the user to do "our" job.

And since this issue is closed :smile_cat:  I should point out that actually the udev replay in the rc.sysinit is probably the slowest step in the boot process...

**Comment by jamesbond3142** [comment]
The sfslist sounds like a good idea.

The only SFS to be loaded by is the basesfs, and probably zdrv (=kernel modules/firmware). The rest of them can be loaded by rc.sysinit, using sfs-load. Why keep the duplicate the code in initrd? sfs-load is powerful and versatile. 

As far as I can see, the reason why we need the rest of the *drv options is because puppy load sfs at bottom of the stack. So an SFS loaded later can't override earlier SFS. The *drv options on the other hand load SFS at top of stack, thus they can overwrite basesfs.

If this is the problem, why not make sfs-load to load at top of stack. Thus the *drv jobs can be done by sfs-load. If searching is needed (e.g. if we want to load sfs files located at the same location of basesfs), let initrd does the searching and keep the result in PUPSTATE; then rc.sysinit can simply refer to this special path and load the rest of the sfs using sfs-load.

We can implement this while still making *drv options backward compatible. Instead of processsing them in initrd, collect all these *drv values in rc.sysinit and pass them to the same function that process sfslist.

**Comment by mavrothal** [comment]
Actually the a/z/y/f-drv they where generated only for building versatility and they all load together as one aufs branch in pup_ro2. Z is for the kernel, A for (optional) applications, Y for alternative desktops (LXDE, XFCE etc) and F for extra firmware. Is pretty much like breaking the original puppy.sfs in pieces to provide different puppy flavours with minimal effort and disruption.
Everything other SFS is taken care by sfs_load after switch root.
Regarding the load order there are obvious advantages loading branches on top. 
The major disadvantage I see is the high risk of messing the system with a bad sfs clobbering needed files. In a well controlled OS this is not a problem, but this is puppy land...
